### PR TITLE
Migrate CI to CircleCI v2, massively reducing CI time

### PR DESCRIPTION
**Pull request by fmoessbauer**
As CircleCI v1 is shut down in August, we have to migrate to v2.0. This brings the possibility to execute all targets / jobs in parallel, which reduces the overall CI time to ~8 minutes.

Another benefit is that the artifacts are now collected even if the job fails, hence it is possible to inspect the logs directly on Circle.

### CODE REVIEW

File: .circleci/config.yml
Comment (devreal) [inline .circleci/config.yml]: Can we set `DASH_MAX_UNITS` to 4? My gut tells me that there might be a) test that require 4 units, and b) some problems don't trigger with less than 4 units. Both are just feelings, though.

File: dash/scripts/circleci/collect-artifacts.sh
Comment (devreal) [inline dash/scripts/circleci/collect-artifacts.sh]: I suggest to move the output under the respective operation

**Review by devreal — CHANGES_REQUESTED** [review]
Thanks for taking care of this :+1: 

Out of curiosity: does this branch already run the CircleCI2 now?

File: .circleci/config.yml
Comment (fmoessbauer) [reply .circleci/config.yml]: Let's try that. In circle v1 this was not possible as the tests run incredibly slow.

File: dash/scripts/circleci/collect-artifacts.sh
Comment (fmoessbauer) [reply dash/scripts/circleci/collect-artifacts.sh]: Yes, done!

**Comment by fmoessbauer** [comment]
> does this branch already run the CircleCI2 now?
Yes, as you can see in the GitHub CI section where now 4 Circle targets are shown.

Btw: I added an additional openmpi3 container now. This was already on my list, but the container has not been marked for automatic builds on dockerhub. ...And building the container takes ~4 hours on there.

File: .circleci/config.yml
Comment (fmoessbauer) [reply .circleci/config.yml]: According to the artifacts, the topology stuff fails on CI. We should either disable this or run with just 3 units. Let's discuss this in the meeting.

**Comment by fmoessbauer** [comment]
The Locality tests still do not run properly on CI, hence I disabled them. This is no regression compared to circle v1, as these tests are only active if running with at least 4 units.

After we have merged this PR, we have to re-run CI on all previous PRs as the merge-guards have changed as well.

**Comment by rkowalewski** [comment]
Before merging we should fix and reenable the `TeamLocalityTest`. Otherwise it will sink into oblivion.

**Comment by fmoessbauer** [comment]
@rkowalewski If @fuchsto is willing to do that, no problem. This is a internal problem in DART locality which does not support a mapping of multiple units to the same core. Even the `--map-by core` in CI seems to not ensure this.

Btw: In the old CI setup this testcase is not executed either, as the test is skipped if running with less than 4 units. So IMO the pragmatic way is to just merge this PR and enable the test after @fuchsto has fixed the bug.

**Comment by devreal** [comment]
I agree with @fmoessbauer, as this is a bug in DART locality it should be handled separately. We cannot tie progress on so many fronts to broken implementations that won't be fixed any time soon...

**Review by rkowalewski — APPROVED** [review]
The changes look fine for me. I am still not lucky with skipped test cases however we should not block this improvement, I agree here.

**Pull request merged**
